### PR TITLE
Update Graph SetInputs and SetOutputs for training

### PR DESCRIPTION
**Pull request by edgchen1**
**Description**
Fix training modification of Graph SetInputs() and SetOutputs(). Originally there were distinct code paths in Graph based on whether the graph was loaded from a GraphProto or created from scratch. The training modifications made that distinction a bit ambiguous - i.e., even though the Graph is loaded from a GraphProto for training, sometimes we rely on the other code path, e.g., to deduce the graph inputs after modifying it. Consequently, there was some odd behavior when using SetInputs(). For correctness, this change separates the cases where the graph is loaded from a GraphProto and where it is created from scratch.

**Motivation and Context**
- Address PR comment: https://github.com/microsoft/onnxruntime/pull/3174#discussion_r397042400
- Fix odd behavior in Graph::SetInputs()

### CODE REVIEW

File: onnxruntime/core/graph/graph.cc [hunk 1]
```diff
@@ -2088,7 +2166,7 @@ void Graph::AddInitializedTensor(const TensorProto& tensor) {
   *(tensor_added) = tensor;
   name_to_initial_tensor_[tensor.name()] = tensor_added;
 
-  if (!GraphLoadedFromModelFile(graph_proto_) && GetNodeArg(tensor.name()) == nullptr) {
+  if (!is_loaded_from_model_file_ && GetNodeArg(tensor.name()) == nullptr) {
     // make sure there is a NodeArg for the initializer as SetGraphInputsOutputs may add it to the graph inputs.
```
Comment (skottmckay) [inline onnxruntime/core/graph/graph.cc:2170]: Probably don't want this first condition any more as we're allowing adding nodes/inputs to a model loaded from proto so the need to create a node arg may exist.

File: onnxruntime/core/graph/graph.cc [hunk 2]
```diff
@@ -2869,13 +2870,33 @@ Status Graph::InlineFunction(Node& node) {
...
+  GraphResolveNeeded(true);
 }
 
 void Graph::SetOutputs(const std::vector<const NodeArg*>& outputs) {
   graph_outputs_ = outputs;
+
   graph_outputs_manually_set_ = true;
+  GraphProtoSyncNeeded(true);
+  GraphResolveNeeded(true); // TODO not sure if this is needed
 }
```
Comment (skottmckay) [inline onnxruntime/core/graph/graph.cc:2900]: I think it is as the resolve should make sure the outputs are produced.

**Comment by skottmckay** [comment]
> 

Could this be inside the 'if', with ComputeOverrideableInitializers also called in SetInputs if loaded from file is true? 

---
Refers to: onnxruntime/core/graph/graph.cc:2706 in 94aead4. [](commit_id = 94aead43c88bbb1e71255df6fe63d3f4cffc5d96, deletion_comment = False)

File: onnxruntime/core/graph/graph.cc [hunk 3]
```diff
@@ -1060,7 +1138,7 @@ Status Graph::BuildConnections(std::unordered_set<std::string>& outer_scope_node
             // If this Graph was built manually, remove the implicit input from the graph outputs if it is present there
             // and not explicitly listed in the ordered graph outputs (as that implies we should leave it as an output).
             // If the Graph was loaded from a GraphProto, honor the explicit graph outputs and leave as is.
-            if (!loaded_from_model_file) {
+            if (!is_loaded_from_model_file_) { // TODO do we need to account for graph_outputs_manually_set_ here?
               graph_outputs_.erase(std::remove(graph_outputs_.begin(), graph_outputs_.end(), node_arg),
```
Comment (skottmckay) [inline onnxruntime/core/graph/graph.cc:1142]: No. SetGraphInputsOutputs is called by Resolve before BuildConnections so they've already been processed.

File: onnxruntime/test/ir/graph_test.cc [hunk 4]
```diff
@@ -1283,5 +1283,49 @@ TEST_F(GraphTest, AddRemoveInitializerHandling) {
...
+
+  graph.AddNode("add_z", "Add", "add z to sum", {sum, z}, {sum_with_z});
+
+  auto inputs = graph.GetInputsIncludingInitializers();
+  inputs.emplace_back(z);
+  graph.SetInputs(inputs);
+
+  auto outputs = graph.GetOutputs();
+  outputs.emplace_back(sum_with_z);
+  graph.SetOutputs(outputs);
```
Comment (skottmckay) [inline onnxruntime/test/ir/graph_test.cc:1319]: nit: prefer push_back unless you really need emplace functionality. [https://abseil.io/tips/112](https://abseil.io/tips/112)

File: orttraining/orttraining/core/graph/graph_augmenter.cc [hunk 5]
```diff
@@ -40,45 +75,34 @@ Status GraphAugmenter::AugmentGraph(Graph& graph, const GraphDefs& graph_element
...
-  }
+  std::vector<const NodeArg*> new_outputs;
+  ORT_RETURN_IF_ERROR(AddToExistingNodeArgs(
+      "override graph outputs", graph, graph_outputs, {}, true, new_outputs));
+  graph.SetOutputs(new_outputs);
 
-  graph.SetOutputs(new_output_args);  // By setting this, Graph::SetGraphInputsOutputs could infer the output as expected.
   graph.SetGraphResolveNeeded();
-  graph.SetGraphProtoSyncNeeded();
   return graph.Resolve();
```
Comment (skottmckay) [inline orttraining/orttraining/core/graph/graph_augmenter.cc:106]: SetOutputs does this now

File: orttraining/orttraining/core/graph/graph_augmenter.cc [hunk 6]
```diff
@@ -40,45 +75,34 @@ Status GraphAugmenter::AugmentGraph(Graph& graph, const GraphDefs& graph_element
...
-  {
-    std::unordered_set<std::string> unique_graph_outputs(graph_outputs.begin(), graph_outputs.end());
-    if (unique_graph_outputs.size() != graph_outputs.size()) {
-      std::string error_message{"The specified graph outputs are not unique:"};
-      for (const auto& graph_output : graph_outputs) {
-        error_message += "\n  " + graph_output;
-      }
-      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, error_message);
-    }
-  }
```
Comment (skottmckay) [inline orttraining/orttraining/core/graph/graph_augmenter.cc:70]: Is this check no longer needed? Having duplicate names in the output can cause issues

**Comment by skottmckay** [comment]
> 

Would probably also need to be called from InitializeStateFromModelFileGraphProto. 

---
In reply to: [610672225](https://github.com/microsoft/onnxruntime/pull/3446#issuecomment-610672225) [](ancestors = 610672225)

---
Refers to: onnxruntime/core/graph/graph.cc:2706 in 94aead4. [](commit_id = 94aead43c88bbb1e71255df6fe63d3f4cffc5d96, deletion_comment = False)

File: onnxruntime/test/ir/graph_test.cc (hunk 4 above)
Comment (edgchen1) [reply onnxruntime/test/ir/graph_test.cc:1319]: interesting point, hadn't considered that emplace could be more error-prone that way.

---
In reply to: [405177466](https://github.com/microsoft/onnxruntime/pull/3446#discussion_r405177466) [](ancestors = 405177466)

File: onnxruntime/core/graph/graph.cc (hunk 1 above)
Comment (edgchen1) [reply onnxruntime/core/graph/graph.cc:2165]: tried to remove the `!is_loaded_from_model_file` but this test started failing: GraphTransformationTests.EmbedLayerNormFusionFormat5

---
In reply to: [405174439](https://github.com/microsoft/onnxruntime/pull/3446#discussion_r405174439) [](ancestors = 405174439)

Comment (edgchen1) [reply onnxruntime/core/graph/graph.cc:2165]: there's a call to graph_utils::AddInitializer() which calls Graph::AddInitializedTensor() prior to calling Graph::GetOrCreateNodeArg(). The TypeProto* passed to the last GetOrCreateNodeArg() contains shape information, but that at that point the NodeArg already exists. So it doesn't get the shape set.

---
In reply to: [405762855](https://github.com/microsoft/onnxruntime/pull/3446#discussion_r405762855) [](ancestors = 405762855,405174439)

File: orttraining/orttraining/test/graph/gradient_graph_builder_test.cc [hunk 7]
```diff
@@ -680,26 +714,11 @@ class PipelineSplitter {
                         name,
                         input_name)) {
         ASSERT_TRUE(is_first);
-        // add shape info
-        EXPECT_TRUE(AddItemByName(sub_gp->mutable_value_info(),
-                                  main_gp.input(),
-                                  name,
-                                  name));
```
Comment (KeDengMS) [inline orttraining/orttraining/test/graph/gradient_graph_builder_test.cc:687]: why remove this? it makes Netron view better...

Comment (edgchen1) [reply orttraining/orttraining/test/graph/gradient_graph_builder_test.cc:687]: i can add them back. on the other hand, in general, we can't rely on value_info being present either.

---
In reply to: [406528154](https://github.com/microsoft/onnxruntime/pull/3446#discussion_r406528154) [](ancestors = 406528154)

File: onnxruntime/core/graph/graph.cc (hunk 1 above)
Comment (edgchen1) [reply onnxruntime/core/graph/graph.cc:2165]: i'll leave this as is for now to minimize changes to Graph

---
In reply to: [405764968](https://github.com/microsoft/onnxruntime/pull/3446#discussion_r405764968) [](ancestors = 405764968,405762855,405174439)

**Pull request merged**